src/main/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapper.java

```diff
@@ -0,0 +1,265 @@
+ *       its implementation logic uses {@link PathMappers#create(Action, OutputPathsMode)} as
+ *       described in its javadocs and has its mnemonic listed in {@link
+ *       PathMappers#SUPPORTED_MNEMONICS}. Such an action must pass the {@link PathMapper} to all
+ *       structured command line constructions. If any unstructured command line arguments refer to
+ *       artifact paths, custom handling needs to be added to {@code mapCustomStarlarkArgv} or
+ *       {@code getMapFn} below.
+ *   <li>A supporting executor strips paths from qualifying actions' inputs and outputs before
+ *       staging for execution by taking {@link Spawn#getPathMapper()} into account.
+ * </ol>
+ *
+ * <p>So an action is responsible for declaring that it strips paths and adjusting its command line
+ * accordingly. The executor is responsible for remapping action inputs and outputs to match.
+ */
+public final class StrippingPathMapper {
+  static final String GUID = "8eb2ad5a-85d4-435b-858f-5c192e91997d";
+
+  /**
+   * Creates a new {@link PathMapper} that strips config prefixes if the particular action instance
+   * supports it.
+   *
+   * @param action the action to potentially strip paths from
+   * @return a {@link StrippingPathMapper} if the action supports it, else {@link Optional#empty()}.
+   */
+  static Optional<PathMapper> tryCreate(Action action) {
+    // This is expected to always be "bazel-out", but we don't want to hardcode it here.
+    PathFragment outputRoot = action.getPrimaryOutput().getExecPath().subFragment(0, 1);
+    if (isPathStrippable(action.getInputs(), outputRoot)) {
+      return Optional.of(
+          create(action.getMnemonic(), action instanceof StarlarkAction, outputRoot));
+    }
+    return Optional.empty();
+  }
+
+  private static PathMapper create(String mnemonic, boolean isStarlarkAction,
+      PathFragment outputRoot) {
+    final StringStripper argStripper = new StringStripper(outputRoot.getPathString());
+    final MapFn<Object> structuredArgStripper = (object, args) -> {
+      if (object instanceof String) {
+        args.accept(argStripper.strip((String) object));
+      } else {
+        args.accept(CommandLineItem.expandToCommandLine(object));
+      }
+    };
+    // This kind of special handling should not be extended. It is a hack that works around a
+    // limitation of the native implementation of location expansion: The output is just a list of
+    // strings, not a structured command line that would allow transparent path mapping.
+    // Instead, reimplement location expansion in Starlark and have it return an Args object.
+    final boolean isJavaAction =
+        mnemonic.equals("Javac") || mnemonic.equals("JavacTurbine") || mnemonic.equals("Turbine");
+    return new PathMapper() {
+      @Override
+      public String getMappedExecPathString(ActionInput artifact) {
+        if (artifact instanceof DerivedArtifact) {
```

Comment:
I was able to find out why some of the internal tests weren't working. It turns out some inputs to the executor that are type ParamFileActionInput were not mapped. Trying out the below seems to work

```
if (artifact instanceof DerivedArtifact || artifact instanceof ParamFileActionInput)
```

I'm suspecting this is because `PathMapper.NOOP` is being passed in `ParameterFileWriteAction` related to [ this comment](https://github.com/bazelbuild/bazel/pull/18098#issuecomment-1559997589)

The other issue that caused test failing was the internal logic preparing the inputs for the executor ws not using the correct path mapper.

Reply:
Thanks for the investigation!

Stripping `ParamFileActionInput` does appear correct to me, essentially that `if` just wants to say "everything but source files" anyway.